src/facade/dragonfly_connection.cc

```diff
@@ -745,33 +749,65 @@ void Connection::DispatchFiber(util::FiberSocketBase* peer) {
   DispatchOperations dispatch_op{builder, this};
   uint64_t request_cache_limit = absl::GetFlag(FLAGS_request_cache_limit);
 
+  size_t squashing_threshold = absl::GetFlag(FLAGS_pipeline_squash);
+
   while (!builder->GetError()) {
     evc_.await(
         [this] { return cc_->conn_closing || (!dispatch_q_.empty() && !cc_->sync_dispatch); });
     if (cc_->conn_closing)
       break;
 
-    MessageHandle msg = move(dispatch_q_.front());
-    dispatch_q_.pop_front();
+    auto recycle = [this, request_cache_limit](MessageHandle msg) {
+      dispatch_q_bytes_.fetch_sub(msg.UsedMemory(), memory_order_relaxed);
 
-    builder->SetBatchMode(dispatch_q_.size() > 0);
+      // Retain pipeline message in pool.
+      if (auto* pipe = get_if<PipelineMessagePtr>(&msg.handle); pipe) {
+        dispatch_q_cmds_count_--;
+        if (stats_->pipeline_cache_capacity < request_cache_limit) {
+          stats_->pipeline_cache_capacity += (*pipe)->StorageCapacity();
+          pipeline_req_pool_.push_back(move(*pipe));
+        }
+      }
+    };
+
+    builder->SetBatchMode(dispatch_q_.size() > 1);
+
+    // Special case: if the dispatch queue accumulated a big number of commands,
+    // we can try to squash them
+    if (dispatch_q_cmds_count_ > squashing_threshold &&
+        dispatch_q_cmds_count_ == dispatch_q_.size()) {
+      vector<CmdArgList> args;
+      args.reserve(dispatch_q_.size());
```

Comment:
Can't it be that the `dispatch_q_` size changed between the if statement above and the current statement? From what I understand the `DispatchFiber` acts as a consumer of the `dispatch_q_` which implies that the producer side `SendAsync` can push to the queue at any time

Reply:
ohhh yes yes!